src/cc/usdt.cc

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2016 GitHub, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <sstream>
+
+#include <fcntl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "bcc_elf.h"
+#include "bcc_proc.h"
+#include "usdt.h"
+#include "vendor/tinyformat.hpp"
+
+namespace USDT {
+
+Probe::Location::Location(uint64_t addr, const char *arg_fmt) : address_(addr) {
+  ArgumentParser_x64 parser(arg_fmt);
+  while (!parser.done()) {
+    Argument *arg = new Argument();
+    if (!parser.parse(arg)) {
+      delete arg;  // TODO: report error
+      continue;
+    }
+    arguments_.push_back(arg);
+  }
+}
+
+Probe::Probe(const char *bin_path, const char *provider, const char *name,
+             uint64_t semaphore)
+    : bin_path_(bin_path),
+      provider_(provider),
+      name_(name),
+      semaphore_(semaphore) {}
+
+bool Probe::in_shared_object() {
+  if (!in_shared_object_)
+    in_shared_object_ = (bcc_elf_is_shared_obj(bin_path_.c_str()) == 1);
+  return in_shared_object_.value();
+}
+
+bool Probe::lookup_semaphore_addr(uint64_t *address, int pid) {
+  auto it = semaphores_.find(pid);
+  if (it != semaphores_.end()) {
+    *address = it->second;
+    return true;
+  }
+
+  if (in_shared_object()) {
+    uint64_t load_address = 0x0;  // TODO
+    *address = load_address + semaphore_;
+  } else {
+    *address = semaphore_;
+  }
+
+  semaphores_[pid] = *address;
+  return true;
+}
+
+bool Probe::add_to_semaphore(int pid, int16_t val) {
+  uint64_t address;
+  if (!lookup_semaphore_addr(&address, pid))
+    return false;
+
+  std::string procmem = tfm::format("/proc/%d/mem", pid);
+  int memfd = ::open(procmem.c_str(), O_RDWR);
+  if (memfd < 0)
+    return false;
+
+  int16_t original;  // TODO: should this be unsigned?
+
+  if (::lseek(memfd, static_cast<off_t>(address), SEEK_SET) < 0 ||
+      ::read(memfd, &original, 2) != 2) {
+    ::close(memfd);
+    return false;
+  }
+
+  original = original + val;
+
+  if (::lseek(memfd, static_cast<off_t>(address), SEEK_SET) < 0 ||
+      ::write(memfd, &original, 2) != 2) {
+    ::close(memfd);
+    return false;
+  }
+
+  ::close(memfd);
+  return true;
+}
+
+bool Probe::enable(int pid) {
+  if (!add_to_semaphore(pid, +1))
+    return false;
+
+  // TODO: what happens if we enable this twice?
+  enabled_semaphores_.emplace(pid, std::move(ProcStat(pid)));
```

Comment:
Clang is warning here: `moving a temporary object prevents copy elision [-Wpessimizing-move]`.

Calling `std::move` here doesn't make sense, since the object created by `ProcStat(pid)` is already an R-Value.